File: src/marker.ts
/** Completion marker utilities shared by daemon and stop-hook. */

export const MARKER_PREFIX = "JACKOPS_TASK_COMPLETE:";

export function completionMarker(taskId: string): string {
  return `${MARKER_PREFIX}${taskId}`;
}

/** Strip common formatting noise that LLMs wrap around markers. */
export function stripFormatting(line: string): string {
  let s = line.trim();
  // Strip markdown code fences
  if (s.startsWith("```") && s.endsWith("```")) {
    s = s.slice(3, -3).trim();
  } else if (s.startsWith("```")) {
    s = s.slice(3).trim();
  } else if (s.endsWith("```")) {
    s = s.slice(0, -3).trim();
  }
  // Strip backticks
  if (s.startsWith("`") && s.endsWith("`")) s = s.slice(1, -1).trim();
  // Strip quotes
  if (
    (s.startsWith('"') && s.endsWith('"')) ||
    (s.startsWith("'") && s.endsWith("'"))
  ) {
    s = s.slice(1, -1).trim();
  }
  // Strip trailing punctuation
  if (s.endsWith(".")) s = s.slice(0, -1).trim();
  return s;
}

/** Check if text contains the completion marker for the given task. */
export function hasCompletionMarker(text: string, taskId: string): boolean {
  const expected = `${MARKER_PREFIX}${taskId}`;
  const lines = text.split("\n");

  // Single pass: scan last 10 non-empty lines from the end.
  // Try cheap includes() first, then expensive stripFormatting() on first 5.
  let checked = 0;
  for (let i = lines.length - 1; i >= 0 && checked < 10; i--) {
    const trimmed = lines[i].trim();
    if (!trimmed) continue;
    checked++;
    if (trimmed.includes(expected)) return true;
    if (checked <= 5 && stripFormatting(trimmed) === expected) return true;
  }

  return false;
}

File: src/marker/marker.go
package marker

import "strings"

const Prefix = "JACKIN_TASK_COMPLETE:"

// CompletionMarker returns the marker string for a given task ID.
func CompletionMarker(taskID string) string {
	return Prefix + taskID
}

// StripFormatting removes markdown code fences, backticks, quotes, and
// trailing punctuation that LLMs sometimes wrap around markers.
func StripFormatting(line string) string {
	s := strings.TrimSpace(line)
	// Strip markdown code fences
	if strings.HasPrefix(s, "```") && strings.HasSuffix(s, "```") {
		s = strings.TrimSpace(s[3 : len(s)-3])
	} else if strings.HasPrefix(s, "```") {
		s = strings.TrimSpace(s[3:])
	} else if strings.HasSuffix(s, "```") {
		s = strings.TrimSpace(s[:len(s)-3])
	}
	// Strip backticks
	if len(s) >= 2 && s[0] == '`' && s[len(s)-1] == '`' {
		s = strings.TrimSpace(s[1 : len(s)-1])
	}
	// Strip quotes
	if len(s) >= 2 && ((s[0] == '"' && s[len(s)-1] == '"') || (s[0] == '\'' && s[len(s)-1] == '\'')) {
		s = strings.TrimSpace(s[1 : len(s)-1])
	}
	// Strip trailing period
	if strings.HasSuffix(s, ".") {
		s = strings.TrimSpace(s[:len(s)-1])
	}
	return s
}

// HasCompletionMarker checks if text contains the completion marker for
// the given task. Scans the last 10 non-empty lines; tries cheap contains
// first, then expensive StripFormatting on the first 5.
func HasCompletionMarker(text, taskID string) bool {
	expected := Prefix + taskID
	lines := strings.Split(text, "\n")

	checked := 0
	for i := len(lines) - 1; i >= 0 && checked < 10; i-- {
		trimmed := strings.TrimSpace(lines[i])
		if trimmed == "" {
			continue
		}
		checked++
		if strings.Contains(trimmed, expected) {
			return true
		}
		if checked <= 5 && StripFormatting(trimmed) == expected {
			return true
		}
	}
	return false
}
